File: tests/issue-26-conformance/check.py
"""Disposable issue #26 Python comparison; not a production validator."""

import json
from pathlib import Path

from jsonschema import Draft202012Validator

HERE = Path(__file__).parent
CASES = json.loads((HERE / "cases.json").read_text(encoding="utf-8"))
SCHEMA = json.loads((HERE / "candidate.schema.json").read_text(encoding="utf-8"))
ZOD_SCHEMA = json.loads((HERE / "zod-generated.schema.json").read_text(encoding="utf-8"))


def manual(value: object) -> bool:
    if not isinstance(value, dict) or set(value) != {"schemaVersion", "instances"}:
        return False
    if value["schemaVersion"] != "1.0":
        return False
    instances = value["instances"]
    if not isinstance(instances, list) or not 1 <= len(instances) <= 3:
        return False
    expected = {"id", "x", "y", "width", "height", "roadRequirement"}
    for item in instances:
        if not isinstance(item, dict) or set(item) != expected:
            return False
        name = item["id"]
        if not isinstance(name, str) or not 1 <= len(name) <= 16:
            return False
        for key, lo, hi in (("x", 0, 63), ("y", 0, 63), ("width", 1, 8), ("height", 1, 8)):
            number = item[key]
            if type(number) is not int or not lo <= number <= hi:
                return False
        if item["roadRequirement"] not in ("NONE", "SINGLE", "DOUBLE"):
            return False
    return True


Draft202012Validator.check_schema(SCHEMA)
Draft202012Validator.check_schema(ZOD_SCHEMA)
schema_first = Draft202012Validator(SCHEMA)
ts_first = Draft202012Validator(ZOD_SCHEMA)
for case in CASES:
    results = {
        "schemaFirstPython": schema_first.is_valid(case["value"]),
        "tsFirstGeneratedPython": ts_first.is_valid(case["value"]),
        "manualPython": manual(case["value"]),
    }
    assert all(result == case["valid"] for result in results.values()), (case["name"], results)
    print(f"{case['name']}: {results}")

File: tests/issue-26-conformance/check.mjs
// Disposable issue #26 comparison. This is not a published snapshot contract.
import assert from "node:assert/strict";
import { readFileSync, writeFileSync } from "node:fs";
import Ajv2020 from "ajv/dist/2020.js";
import { compile } from "json-schema-to-typescript";
import * as z from "zod";

const schema = JSON.parse(
  readFileSync(new URL("candidate.schema.json", import.meta.url)),
);
const cases = JSON.parse(readFileSync(new URL("cases.json", import.meta.url)));
const ajv = new Ajv2020({ strict: true, allErrors: true });
const schemaValidate = ajv.compile(schema);

const item = z.strictObject({
  id: z.string().min(1).max(16),
  x: z.int().min(0).max(63),
  y: z.int().min(0).max(63),
  width: z.int().min(1).max(8),
  height: z.int().min(1).max(8),
  roadRequirement: z.enum(["NONE", "SINGLE", "DOUBLE"]),
});
const tsFirst = z.strictObject({
  schemaVersion: z.literal("1.0"),
  instances: z.array(item).min(1).max(3),
});
const generatedFromTs = z.toJSONSchema(tsFirst);
writeFileSync(
  new URL("zod-generated.schema.json", import.meta.url),
  JSON.stringify(generatedFromTs, null, 2) + "\n",
);
const types = await compile(schema, "ExperimentalCityShape", {
  bannerComment: "",
});
assert.match(types, /roadRequirement: "NONE" \| "SINGLE" \| "DOUBLE"/);
assert.match(types, /schemaVersion: "1\.0"/);
writeFileSync(new URL("generated.d.ts", import.meta.url), types);

const keys = (value, expected) =>
  value &&
  typeof value === "object" &&
  !Array.isArray(value) &&
  Object.keys(value).sort().join("|") === expected.slice().sort().join("|");
const integer = (value, min, max) =>
  Number.isInteger(value) && value >= min && value <= max;
const manual = (value) =>
  keys(value, ["schemaVersion", "instances"]) &&
  value.schemaVersion === "1.0" &&
  Array.isArray(value.instances) &&
  value.instances.length >= 1 &&
  value.instances.length <= 3 &&
  value.instances.every(
    (v) =>
      keys(v, ["id", "x", "y", "width", "height", "roadRequirement"]) &&
      typeof v.id === "string" &&
      v.id.length >= 1 &&
      v.id.length <= 16 &&
      integer(v.x, 0, 63) &&
      integer(v.y, 0, 63) &&
      integer(v.width, 1, 8) &&
      integer(v.height, 1, 8) &&
      ["NONE", "SINGLE", "DOUBLE"].includes(v.roadRequirement),
  );

for (const c of cases) {
  const results = {
    schemaFirstAjv: schemaValidate(c.value),
    tsFirstZod: tsFirst.safeParse(c.value).success,
    manualJs: manual(c.value),
  };
  assert.deepEqual(
    results,
    Object.fromEntries(Object.keys(results).map((k) => [k, c.valid])),
    c.name,
  );
  console.log(`${c.name}: ${JSON.stringify(results)}`);
}
console.log(`Generated TypeScript declaration: ${types.length} characters`);
